contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -2689,6 +2699,169 @@ pltsql_detect_numeric_overflow(int weight, int dscale, int first_block, int nume
 	return (total_digit_count > TDS_NUMERIC_MAX_PRECISION);
 }
 
+/*
+ * Throw a common error message while casting to time datatype
+ */
+static void gen_func_time_in_error()
+{
+	ereport(ERROR,
+			(errcode(ERRCODE_INVALID_DATETIME_FORMAT),
+			errmsg("Conversion failed when converting date and/or time from character string.")));
+}
+
+static bool containsInMonthFormat(int *ftype){
+	int count0 = 0, count1 = 0;
+	for(int i=0;i<3;i++)
+	{
+		if(ftype[i] == DTK_NUMBER)
+			count0++;
+		else if(ftype[i] == DTK_STRING)
+			/* To-Do Check if valid text month is given */
+			count1++;
+		else
+			return false;
+	}
+	if(count0 ==2 && count1 ==1)
+		return true;
+	return false;
+}
+
+Datum pltsql_time_in(const char *str, int32	typmod)
+{
+	TimeADT		result;
+	fsec_t		fsec;
+	struct pg_tm tt,
+			   *tm = &tt;
+	int			tz;
+	int			nf;
+	int			dterr;
+	char		workbuf[MAXDATELEN + 1];
+	char	   *field[MAXDATEFIELDS];
+	int			dtype;
+	int			ftype[MAXDATEFIELDS];
+	StringInfo	sf_1 ;
+	int status;
+
+	dterr = ParseDateTime(str, workbuf, sizeof(workbuf),
+						field, ftype, MAXDATEFIELDS, &nf);
+	if (dterr == 0)
+	{
+		if(nf >= 3)
+		{
+			if(containsInMonthFormat(ftype))
+			{
+				StringInfoData buf;
+				initStringInfo(&buf);
+				appendStringInfoString(&buf, field[0]);
+				for(int i=1;i<3;i++)
+				{
+					appendStringInfoString(&buf, "-");
+					appendStringInfoString(&buf, field[i]);
+
+				}
+				field[0]=buf.data;
+				ftype[0] = DTK_DATE;
+				nf = nf - 2;
+				for(int i = 1; i < nf; i++)
+				{
+					field[i] = field[i+2];
+					ftype[i] = ftype[i+2];
+				}
+			}
+		}
+
+		for(int i=0;i < nf; i++)
+		{
+			switch(ftype[i])
+			{
+				case DTK_NUMBER:
+					switch(strlen(field[i]))
+					{
+						case 1:
+						case 2:
+							/*
+							 * At this point the 1,2 digit number should be considered as
+							 * time only if there is an [AP]M after the number or else
+							 * should throw an error
+							 */
+							if(i == nf-1 || ftype[i+1] != DTK_STRING)
+							{
+								gen_func_time_in_error();
+							}
+							status = RE_compile_and_execute(cstring_to_text("^([AP]M)$"),
+															(char *) field[i+1], strlen(field[i+1]),
+															REG_ADVANCED | REG_ICASE, DEFAULT_COLLATION_OID,
+															0, NULL);
+							if(!status)
+								gen_func_time_in_error();
+
+							sf_1 = makeStringInfo();
+							appendStringInfo(sf_1, "%s%s", field[i], ":0:0");
+							field[i] = sf_1->data;
```

Comment:
free existing field[i]

Reply:
sure, will do that in next fix.

Reply:
Free of fiel, cannot happen as it will result in crash because of not initializing using malloc/palloc.